first-party/jib-classpath-main-class-extension-gradle/src/main/java/com/google/cloud/tools/jib/gradle/extension/classpathmainclass/ClasspathMainClassExtractor.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2021 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.gradle.extension.classpathmainclass;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+public class ClasspathMainClassExtractor {
+
+  private static List<String> classpathOptions = Arrays.asList("-cp", "-classpath", "--class-path");
+
+  static Map<String, String> extractEnvsFromEntrypoint(@Nullable List<String> entrypoint) {
+    if (entrypoint == null || entrypoint.size() == 0) {
+      return Collections.emptyMap();
+    }
+
+    if (!entrypoint.get(0).equals("java")) {
+      return Collections.emptyMap();
+    }
+
+    int i = entrypoint.size() - 1;
+    final String mainClass = entrypoint.get(i);
+    if (--i < 0) {
+      return Collections.emptyMap();
+    }
+
+    String classpathString = entrypoint.get(i);
+    if (--i < 0) {
+      return Collections.emptyMap();
+    }
+
+    if (!classpathOptions.contains(entrypoint.get(i))) {
+      return Collections.emptyMap();
+    }
```

Comment:
I was talking about

```java
for (int i = ... ) {
  if (classpathOptions.contains(entrypoint.get(i))) {
    classpathString = entrypoint.get(i + 1);
    break;
  }
}
```

Reply:
not sure why you want to loop here. the simple extractor probably only works with the code that creates the entrypoint list is not changed, otherwise you begin to replicate java command line parser here to cover all cases.

Reply:
There's a possibility that we will change the main Jib code to rearrange the entrypoint in the future, and when doing it, it's easy that we forget that we have this extension and the need to maintain the invariant that `-cp` must be the last JVM option. It's easy to break this extension unintentionally. And when that happens, it becomes particularly complicated for us and the user: we need to re-publish a new version of this extension, and the user has to upgrade the extension. We want an extension to be very stable.

It's always about a trade-off. Obviously, replicating a well-behaving a command line parser is an overkill. But it's really trivial to cover this case with a simple loop. Considering the possibility that the entrypoint may change in the future, I really think this is worth.

Reply:
okay, understood, sound sensible.

Reply:
is it okay to keep the assumption that the last argument is the main class? otherwise it's not to easy to distinguish between main class and program arguments.

Reply:
Yeah, I think that's acceptable. I'll be out of office tomorrow, so I'll take another look at this next Monday. Thanks!